compiler-core/src/type_/expression.rs

```diff
@@ -10,28 +10,131 @@ use crate::{
         UntypedExprBitArraySegment, UntypedMultiPattern, UntypedStatement, Use, UseAssignment,
         USE_ASSIGNMENT_VARIABLE,
     },
+    build::Target,
     exhaustiveness,
 };
 use id_arena::Arena;
 use im::hashmap;
 use itertools::Itertools;
 use vec1::Vec1;
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct SupportedTargets {
+    erlang: bool,
+    javascript: bool,
+}
+
+impl SupportedTargets {
+    pub fn none() -> SupportedTargets {
+        SupportedTargets {
+            erlang: false,
+            javascript: false,
+        }
+    }
+
+    pub fn all() -> SupportedTargets {
+        SupportedTargets {
+            erlang: true,
+            javascript: true,
+        }
+    }
+
+    pub fn javascript() -> SupportedTargets {
+        SupportedTargets {
+            erlang: false,
+            javascript: true,
+        }
+    }
+
+    pub fn erlang() -> SupportedTargets {
+        SupportedTargets {
+            erlang: true,
+            javascript: false,
+        }
+    }
+
+    pub fn from_target(target: Target) -> SupportedTargets {
+        match target {
+            Target::Erlang => SupportedTargets::erlang(),
+            Target::JavaScript => SupportedTargets::javascript(),
+        }
+    }
+
+    pub fn intersect(&self, targets: SupportedTargets) -> SupportedTargets {
+        SupportedTargets {
+            erlang: self.erlang && targets.erlang,
+            javascript: self.javascript && targets.javascript,
+        }
+    }
```

Comment:
Perhaps this could be called `narrow` as that's the word used elsewhere in this PR.

Let's also use mutation here.